src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs

```diff
@@ -46,8 +46,9 @@ private void RegisterCommandCallback()
 
         private void OnEventSourceCommand(object? sender, EventCommandEventArgs e)
         {
-            // Should only be enable or disable
-            Debug.Assert(e.Command == EventCommand.Enable || e.Command == EventCommand.Disable);
+            // Should only be enable or disable, but during EventSource initialization, update may be seen as this callback
+            // is invoked on all m_deferredCommands while processing them synchronously through EventSource.DoCommand.
```

Comment:
This comment seems technically correct, but also misleading :) If I understand the situation correctly, the callstack where we see an Update event looks like this:

```
CounterGroup.OnEventSourceCommand
EventSource.add_EventCommandExecuted
CounterGroup.RegisterCommandCallback
CounterGroup.ctor
CounterGroup.GetCounterGroup
DiagnosticCounter.Publish
PollingCounter.ctor
EventSource.OnEventCommand
EventSource.DoCommand
EventSource.Initialize
```

It is technically true that the event was received while executing EventSource.DoCommand(), but I bet most people reading the comment would assume the call is coming from [here](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs,2704). Instead it is coming via this very round about call sequence.

If we leave this as-is I'd recommend clarifying in the comment about the subtle call sequence, but I think it would be better to fix the logic in EventSource that handles defered command notification. Looking at the code I think there are several potential issues:
1. The callback from EventSource.add_EventCommandExecuted can deliver different command args than the callback from DoCommand(). DoCommand() modifies Update events when it runs but add_EventCommandExecuted might dispatch them before that happens.
2. The callback from EventSource.add_EventCommandExecuted can also cause defered commands to generate two callback invocations because DoCommand() will duplicate the callback later. Not necessarily a bug, but unexpected.
3. Setting m_completelyInitialized=true [here](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs,1642) means that command which arrives before Initialize() enters [this lock](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs,1651) would cause all deferred commands to be [deleted](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs,2539) before we gave any notification.

I'm guessing the weird logic with m_completelyInitialized and waiting to receive another command before getting rid of the deferred commands is to let people write code like this and have their callback observe the deferred commands:
```C#
MyEventSource es = new MyEventSource();
es.EventCommandExecuted += m_callback;
```

That code already isn't guaranteed to work in the presence of a concurrently received event, but most scenarios probably don't have such an event so it turns out OK. 

I think most of weird behavior (including this bug) could be resolved by moving m_completelyInitialized = true to occur after we've dispatched the deferred commands:
```C#
lock (EventListener.EventListenersLock)
{
    EventCommandEventArgs? deferredCommands = m_deferredCommands;
    while (deferredCommands != null)
    {
        DoCommand(deferredCommands);
        deferredCommands = deferredCommands.nextCommand;
    }
    m_completelyInitialized = true;
}
```

Then we can make callback inside EventCommandExecuted conditional so that it only activates after m_completelyInitialized==true

```C#
                // If we have an EventHandler<EventCommandEventArgs> attached to the EventSource before the first command arrives
                // It should get a chance to handle the deferred commands.
                if(m_completelyInitialized)
                {
                    EventCommandEventArgs? deferredCommands = m_deferredCommands;
                    while (deferredCommands != null)
                    {
                        value(this, deferredCommands);
                        deferredCommands = deferredCommands.nextCommand;
                    }
                }
```

Reply:
Those are very good points! I didn't even realize that there was `this.m_eventCommandExecuted?.Invoke(this, commandArgs);` since I was so focused on tracing the Assert error 😅.

I'm just realizing that EventCommandExecuted allows users to observe deferred commands in case EventSource fails to initialize. If we continue to associate `m_completelyInitialized` with initialization completed without exception (given its [comment](https://github.com/dotnet/runtime/blob/dff1d8467845fd93c517f89dc81598e5fd17c270/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L3822)) by
```C#
// Initialize post try/catch
lock (EventListener.EventListenersLock)
{
    EventCommandEventArgs? deferredCommands = m_deferredCommands;
    while (deferredCommands != null)
    {
        DoCommand(deferredCommands);
        deferredCommands = deferredCommands.nextCommand;
    }
    m_completelyInitialized = m_constructionException == null;
}
```
then guarding inside `EventCommandExecuted`
```C#
                if (m_completelyInitialized)
                {
                    EventCommandEventArgs? deferredCommands = m_deferredCommands;
                    while (deferredCommands != null)
                    {
                        value(this, deferredCommands);
                        deferredCommands = deferredCommands.nextCommand;
                    }
                }
```
will prevent that former capability. If we don't think there is any value in observing deferred commands should EventSource fail to initialize, then that is fine.
(I'm not sure what value there is in being able to see all deferred commands if the EventSource fails to initialize, unless maybe users retry EventSource initialization and requeue those deferred commands, but is EventSource flakey enough to fail initialization one time and succeed the next time?). 

However, if we want to change `m_completelyInitialized` to just mean "initialization completed" regardless of success/exception with
```C#
// Initialize post try/catch
lock (EventListener.EventListenersLock)
{
    EventCommandEventArgs? deferredCommands = m_deferredCommands;
    while (deferredCommands != null)
    {
        DoCommand(deferredCommands);
        deferredCommands = deferredCommands.nextCommand;
    }
    m_completelyInitialized = true;
}
```
then we would probably have to change SendCommand's condition to clear `m_deferredCommands` to `m_completely_initialized && m_constructionException == null` if we want to allow users to observe deferred commands if an initialization exception occurred.


I'm wondering, does `CounterGroup` call [`RegisterCommandCallback`](https://github.com/dotnet/runtime/blame/dff1d8467845fd93c517f89dc81598e5fd17c270/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs#L25) solely to set a callback for event commands sent after EventSource is completely initialized (I'm thinking this is the case), and that `EventCommandExecuted` invoking the callback for m_deferredCommands is an undesired consequence? Maybe there should be a separation between setting the `EventCommandExecuted` callback and invoking the callback on deferred commands for https://github.com/dotnet/runtime/blob/dff1d8467845fd93c517f89dc81598e5fd17c270/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L503-L504.

Reply:
Implemented the former suggestion, so now `EventCommandExecuted` will not be able to observe deferred_commands if the EventSource fails to initialize.